plugin/code_actions.py

```diff
@@ -63,52 +84,215 @@ def get_location_key(self, view: sublime.View, point: int) -> str:
+        actions that need to be filtered.
+        """
+        def actions_filter(actions: CodeActionsResponse) -> List[CodeActionOrCommand]:
+            return [a for a in (actions or []) if a.get('kind') in kinds]
+        collector = actions_collector.create_collector(config_name)
+        return (
+            lambda actions: collector(actions_filter(actions)),
+            lambda error: collector([])
+        )
+
+    actions_collector = CodeActionsCollector(actions_handler)
+    with actions_collector:
+        file_name = view.file_name()
+        if file_name:
+            sessions = sessions_for_view(view, 'codeActionProvider')
+            for session in sessions:
+                supported_kinds = session.get_capability('codeActionProvider.codeActionKinds')
+                matching_kinds = get_matching_kinds(on_save_actions, supported_kinds or [])
+                if matching_kinds:
+                    params = _create_code_action_request_params(view, file_name, [], matching_kinds)
+                    request = Request.codeAction(params)
+                    collect, onerror = filtering_collector(session.config.name, matching_kinds, actions_collector)
+                    session.send_request(request, collect, onerror)
+    return actions_collector
+
+
+def _create_code_action_request_params(
+    view: sublime.View,
+    file_name: str,
+    diagnostics: List[Diagnostic],
+    on_save_actions: Optional[List[str]] = None
+) -> Dict:
+    if on_save_actions:
+        relevant_range = entire_content_range(view)
+    else:
+        relevant_range = diagnostics[0].range if diagnostics else region_to_range(view, view.sel()[0])
+    params = {
+        "textDocument": {
+            "uri": filename_to_uri(file_name)
+        },
+        "range": relevant_range.to_lsp(),
+        "context": {
+            "diagnostics": list(diagnostic.to_lsp() for diagnostic in diagnostics)
+        }
+    }
+    if on_save_actions:
+        params['context']['only'] = on_save_actions
+    return params
+
+
+def get_matching_kinds(user_actions: Dict[str, bool], session_actions: List[str]) -> List[str]:
+    """
+    Filters user-enabled or disabled actions so that only ones matching the session actions
+    are returned. Returned actions are those that are enabled and are not overridden by more
+    specific, disabled actions.
+
+    Filtering only returns actions that exactly match the ones supported by given session.
+    If user has enabled a generic action that matches more specific session action
+    (for example user's a.b matching session's a.b.c), then the more specific (a.b.c) must be
+    returned as servers must receive only actions that they advertise support for.
+    """
+    matching_kinds = []
+    for session_action in session_actions:
+        enabled = False
+        action_parts = session_action.split('.')
+        for i in range(len(action_parts)):
+            current_part = '.'.join(action_parts[0:i + 1])
```

Comment:
This looks like an O(N^2) algorithm?

Can we instead iterate over the parts directly?
```
keys = session_action.split('.')
for key in keys:
   ...

Reply:
We need index to be able to create a "cut out" of current "parts" (like take "a.b" from "a.b.c").

In any case, I don't think we should be concerned about performance here. This runs on a very small input and only once per session per save command.